### PR TITLE
EDU-670: Backport Java into Python Coverage

### PR DESCRIPTION
**Pull request by fairlydurable**

### CODE REVIEW

File: docs/develop/java/message-passing.mdx [hunk 1]
```diff
@@ -91,17 +95,18 @@ public class MessagePassingIntro {
 }
 ```
 
-- The annotation can take arguments. See the API reference docs: [`@QueryMethod`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/workflow/QueryMethod.html).
-- A Query handler must not modify workflow state
-- You can't do any blocking operations such as executing an Activity in a Query handler.
+- The Query decorator can accept arguments.
```
Comment (dandavison) [inline docs/develop/java/message-passing.mdx:98]: Note that what Python calls a "decorator" is called an "annotation" in Java.

File: docs/develop/java/message-passing.mdx [hunk 2]
```diff
@@ -33,16 +33,20 @@ Temporal Clients use messages to read Workflow state and control execution.
...
 
-General guidelines for writing message handlers:
+Follow these guidelines when writing your message handlers:
 
-- Arguments and return values of handlers (and the main Workflow function) must be [serializable](/dataconversion): a class will often be the right choice.
-- It's possible to write handler methods that take multiple arguments, but this is not recommended: instead use a single class argument to which fields can be added as needed.
+- Message handlers are defined as methods on the Workflow class, using one of the three decorators: [`@workflow.query`](https://python.temporal.io/temporalio.workflow.html#query), [`@workflow.signal`](https://python.temporal.io/temporalio.workflow.html#signal), and [`@workflow.update`](https://python.temporal.io/temporalio.workflow.html#update).
+- The parameters and return values of handlers and the main Workflow function must be [serializable](/dataconversion).
+- Prefer [data classes](https://docs.python.org/3/library/dataclasses.html) to multiple input parameters.
+  Data class parameters allow you to add fields without changing the calling signature.
```
Comment (dandavison) [inline docs/develop/java/message-passing.mdx:47]: Just leaving a TODO here for conversion to Java

Comment (fairlydurable) [reply docs/develop/java/message-passing.mdx:47]: D'erp again. Can't believe I missed this. Fixing.

**Review by dandavison — COMMENTED** [review]
I've merged master and added one commit applying SDK team technical content fixes. I have not yet covered the "Troubleshooting" and subsequent sections.

File: docs/develop/java/message-passing.mdx [hunk 3]
```diff
@@ -289,7 +331,11 @@ public static void signalWithStart() {
 }
 ```
 
-The following example shows the Workflow interface for the `GreetingWorkflow` called in the previous example.
+<div style={{backgroundColor: '#ffff00',padding: '0px 15px',borderRadius: '5px',border: '1px solid #cccccc',display: 'inline-block'}}>**Is this correct?**</div>
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/develop/java/message-passing.mdx]: Getting so excited to see this come together. Let me know when I should be hopping back in.

**Review by dandavison — APPROVED** [review]
LGTM! Just needs the one `URL TODO` to be addressed.

**Comment by dandavison** [comment]
I've deleted the link to the sample (which had a TODO comment) in order for this to be mergable. To reinstate the sample link we need to merge https://github.com/temporalio/samples-java/pull/662.

**Comment by dandavison** [comment]
I've reverted the dynamic components section to `main` and made the "dynamic updates" section match the others. That's sufficient for now; we can update the dynamic components section in future PRs.

File: docs/develop/java/message-passing.mdx [hunk 4]
```diff
@@ -439,142 +581,155 @@ public class DataWorkflowImpl implements DataWorkflow {
...
+- **The handler caused the Workflow Task to fail.**
+  This would happen, for example, if the Query handler blocks the thread for too long without yielding.
 
-        You'll get a [`WorkflowServiceException`](https://javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/client/WorkflowServiceException.html) "workflow execution already completed"`
+## Dynamic components {#dynamic-handler}
 
-        This will happen if the Workflow finished while the Update handler execution was in progress, for example because
-        - The Workflow was canceled or failed.
-        - The Workflow completed normally or continued-as-new and the Workflow author did not [wait for handlers to be finished](/encyclopedia/workflow-message-passing#finishing-message-handlers).
+A dynamic Workflow, Activitity, Signal, Update, or Query is a kind of unnamed item.
```
Comment (fairlydurable) [inline docs/develop/java/message-passing.mdx:622]:
```suggestion
A dynamic Workflow, Activity, Signal, Update, or Query is a kind of unnamed item.
```

**Pull request merged**